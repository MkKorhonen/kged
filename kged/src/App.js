import React from 'react';
import logo from './logo.svg';
import './App.css';

function App() {
  return (
    <div className="App">
      <div className="left">
        A
      </div>
      <div className="center">
        B
      </div>
      <div className="right">
        C
      </div>
    </div>
  );
}

export default App;
