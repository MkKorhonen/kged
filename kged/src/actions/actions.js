export const GET_ROOM = 'GET_ROOM'

export function getRoom(room) {
    return { type: GET_ROOM, room}
}
