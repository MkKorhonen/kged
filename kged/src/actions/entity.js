export const getActiveEntity = (state) => {
    let activeEntityId = state.entity.activeEntity
    return [
        ...state.rooms.rooms,
        ...state.furnitures.furnitures,
        ...state.items.items
    ].find(e => e.attrs.id === activeEntityId)
}

export const setActiveEntity = (id) => ({
    type: 'SET_ACTIVE_ENTITY',
    payload: {
        id: id
    }
})

export const removeActiveEntity = () => ({
    type: 'REMOVE_ACTIVE_ENTITY'
})

