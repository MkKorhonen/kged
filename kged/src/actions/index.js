// action creators

import { fetchRooms, exportRooms, fetchFurnitures } from 'api'

export const setActiveEntity = (entity) => ({
    type: 'SET_ACTIVE_ENTITY',
    payload: {
        entity: entity
    }
})

// thunk
export function loadRooms() {
    return function(dispatch, getState) {
        const rooms = fetchRooms()
        dispatch({
            type: 'GET_ROOMS',
            payload: {
                rooms
            }
        })
    }
}

export const addRoom = (room) => ({
    type: 'ADD_ROOM',
    payload: {
        room: room
    }
})

export const updateRoom = (room) => ({
    type: 'UPDATE_ROOM',
    payload: {
        room: room
    }
})

export const deleteRoom = (room) => ({
    type: 'DELETE_ROOM',
    payload: {
        room: room
    }
})

export function setActiveRoom(room) {
    return function(dispatch) {
        dispatch(setActiveEntity(room))
        dispatch({
            type: 'SET_ACTIVE_ROOM',
            payload: {
                room: room
            }
        })
    }
}

export function saveRooms() {
    return function(dispatch, getState) {
        const state = getState()
        exportRooms(state.rooms.rooms)
        dispatch({
            type: 'EXPORT_ROOMS',
            payload: {}
        })
    }
}




export function loadFurnitures() {
    return function(dispatch, getState) {
        const furnitures = fetchFurnitures()
        dispatch({
            type: 'GET_FURNITURES',
            payload: {
                furnitures
            }
        })
    }
}

export const addFurniture = (furniture) => ({
    type: 'ADD_FURNITURE',
    payload: {
        furniture: furniture
    }
})

export const updateFurniture = (furniture) => ({
    type: 'UPDATE_FURNITURE',
    payload: {
        furniture: furniture
    }
})

export const deleteFurniture = (furniture) => ({
    type: 'DELETE_FURNITURE',
    payload: {
        furniture: furniture
    }
})

export function setActiveFurniture(furniture) {
    return function(dispatch) {
        dispatch(setActiveEntity(furniture))
        dispatch({
            type: 'SET_ACTIVE_FURNITURE',
            payload: {
                furniture: furniture
            }
        })
    }
}

