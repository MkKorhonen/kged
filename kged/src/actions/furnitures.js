import { fetchFurnitures } from 'api'
import { setActiveEntity, updateActiveEntity, removeActiveEntity } from './entity'
import { isExistingEntity } from 'utils'
import { DuplicateEntityError } from 'utils/errors'

export const loadFurnitures = () => {
    return (dispatch, getState) => {
        const furnitures = fetchFurnitures()
        dispatch({
            type: 'GET_FURNITURES',
            payload: {
                furnitures
            }
        })
    }
}

export const setFurnitureImage = (furnitureId, filePath, objectUrl) => {
    return (dispatch, getState) => {
        dispatch({
            type: 'SET_FURNITURE_IMAGE',
            payload: {
                furnitureId: furnitureId,
                filePath: filePath,
                objectUrl: objectUrl
            }
        })
        dispatch(updateActiveFurniture())
    }
}

export const addFurniture = (furniture) => ({
    type: 'ADD_FURNITURE',
    payload: {
        furniture: furniture
    }
})

export const updateFurnitureId = (oldId, newId) => {
    return (dispatch, getState) => {
        if (isExistingEntity(getState(), newId)) {
            throw new DuplicateEntityError('Nimi on jo käytössä')
        }
        dispatch({
            type: 'UPDATE_FURNITURE_ID',
            payload: {
                oldId: oldId,
                newId: newId
            }
        })
        dispatch(updateActiveFurniture(newId))
    }
}

export const deleteFurniture = (furniture) => {
    return (dispatch, getState) => {
        dispatch({
            type: 'DELETE_FURNITURE',
            payload: {
                furniture: furniture
            }
        })
        dispatch(removeActiveEntity())
    }
}

export const setActiveFurniture = (furniture) => {
    return (dispatch) => {
        dispatch(setActiveEntity(furniture))
        dispatch({
            type: 'SET_ACTIVE_FURNITURE',
            payload: {
                furniture: furniture
            }
        })
    }
}

export const updateActiveFurniture = (id = null) => {
    return (dispatch) => {
        dispatch({
            type: 'UPDATE_ACTIVE_FURNITURE',
            payload: { id: id }
        })
        dispatch(updateActiveEntity({category: 'furniture', id: id}))
    }
}
