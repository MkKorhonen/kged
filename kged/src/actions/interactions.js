import { sortBy } from 'lodash/fp'

import { fetchInteractions } from 'api'
import { setActiveEntity, removeActiveEntity } from './entity'
import { isExistingEntity } from 'utils'
import { DuplicateEntityError } from 'utils/errors'
import 'styles/interactions.scss'

export const getInteractions = (state) => {
    return sortBy('attrs.id')(state.interactions.interactions)
}

export const loadInteractions = () => {
    return (dispatch, getState) => {
        const interactions = fetchInteractions()
        dispatch({
            type: 'GET_INTERACTIONS',
            payload: {
                interactions
            }
        })
    }
}

export const addInteraction = (interaction) => ({
    type: 'ADD_INTERACTION',
    payload: {
        interaction: interaction
    }
})

export const updateInteraction = (oldId, newData) => {
    return (dispatch, getState) => {
        if (oldId !== newData.name && isExistingEntity(getState(), newData.name)) {
            throw new DuplicateEntityError('Nimi on jo käytössä')
        }
        dispatch({
            type: 'UPDATE_INTERACTION',
            payload: {
                oldId: oldId,
                newId: newData.name,
                source: newData.source,
                destination: newData.destination,
                interactionType: newData.interactionType,
                command: newData.command
            }
        })
        dispatch(updateActiveInteraction(newData.name))
    }
}

export const deleteInteraction = (interaction) => {
    return (dispatch, getState) => {
        dispatch({
            type: 'DELETE_INTERACTION',
            payload: {
                interaction: interaction
            }
        })
        dispatch(removeActiveEntity())
    }
}

export const setActiveInteraction = (interaction) => {
    return (dispatch) => {
        dispatch(setActiveEntity({id: interaction}))
        dispatch({
            type: 'SET_ACTIVE_INTERACTION',
            payload: {
                interaction: interaction
            }
        })
    }
}

export const updateActiveInteraction = (id = null) => {
    return (dispatch) => {
        dispatch({
            type: 'UPDATE_ACTIVE_INTERACTION',
            payload: { id: id }
        })
    }
}
