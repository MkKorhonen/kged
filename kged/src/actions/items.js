import sortBy from 'lodash/fp/sortBy'

import { fetchItems } from 'api'
import { setActiveEntity, updateActiveEntity, removeActiveEntity } from './entity'
import { isExistingEntity } from 'utils'
import { DuplicateEntityError } from 'utils/errors'

export const getItems = (state) => {
    return sortBy('attrs.id')(state.items.items)
}

export const loadItems = () => {
    return (dispatch, getState) => {
        const items = fetchItems()
        dispatch({
            type: 'GET_ITEMS',
            payload: {
                items
            }
        })
    }
}

export const setItemImage = (itemId, filePath, objectUrl) => {
    return (dispatch, getState) => {
        dispatch({
            type: 'SET_ITEM_IMAGE',
            payload: {
                itemId: itemId,
                filePath: filePath,
                objectUrl: objectUrl
            }
        })
        dispatch(updateActiveItem())
    }
}

export const addItem = (item) => ({
    type: 'ADD_ITEM',
    payload: {
        item: item
    }
})

export const updateItem = (oldId, newData) => {
    return (dispatch, getState) => {
        if (oldId !== newData.name && isExistingEntity(getState(), newData.name)) {
            throw new DuplicateEntityError('Nimi on jo käytössä')
        }
        dispatch({
            type: 'UPDATE_ITEM',
            payload: {
                oldId: oldId,
                newId: newData.name,
                containerRoom: newData.selectedItem.attrs.id,
                xValue: newData.xValue,
                yValue: newData.yValue
            }
        })
        dispatch(updateActiveItem(newData.name))
    }
}

export const deleteItem = (item) => {
    return (dispatch, getState) => {
        dispatch({
            type: 'DELETE_ITEM',
            payload: {
                item: item
            }
        })
        dispatch(removeActiveEntity())
    }
}

export const setActiveItem = (item) => {
    return (dispatch) => {
        dispatch(setActiveEntity(item))
        dispatch({
            type: 'SET_ACTIVE_ITEM',
            payload: {
                item: item
            }
        })
    }
}

export const updateActiveItem = (id = null) => {
    return (dispatch) => {
        dispatch({
            type: 'UPDATE_ACTIVE_ITEM',
            payload: { id: id }
        })
        dispatch(updateActiveEntity({category: 'item', id: id}))
    }
}
