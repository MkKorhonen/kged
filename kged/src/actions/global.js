import * as api from 'api'
import { saveAs } from 'file-saver'
import JSZip from 'jszip'
import { loadRooms } from './rooms'
import { loadItems } from './items'
import { filterFurnitures, extractFurnitures } from 'utils/index'
import { loadFurnitures } from './furnitures'

export const exportProject = (event) => {
    return (dispatch, getState) => {
        const state = getState()
        const rooms = api.exportRooms(state)
        const items = api.exportItems(state)
        const interactions = api.exportInteractions(state)

        const roomsToJSON = JSON.stringify({rooms: rooms}, null, 4)
        const itemsToJSON = JSON.stringify(items, null, 4)
        const interactionsToJSON = JSON.stringify(interactions, null, 4)

        const zip = new JSZip();
        zip.file('rooms.json', roomsToJSON);
        zip.file('items.json', itemsToJSON);
        zip.file('interactions.json', interactionsToJSON);
        zip.generateAsync({type: 'blob'})
        .then(content => {
            saveAs(content, 'game_data.zip');
        });
    }
}

export const importProject = (pkg) => {
    return (dispatch, getState) => {
        JSZip.loadAsync(pkg).then(zip => {
            let files = Object.keys(zip.files)
            let interactionFile = zip.files['interactions.json']
            let interactionData
            (async () => {
                return await interactionFile.async('string').then(data => {interactionData = JSON.parse(data)})
            })()
            files.forEach(filename => {
                zip.files[filename].async('string').then(fileData => {
                    const name = zip.files[filename].name.trim()
                    const data = JSON.parse(fileData)
                    switch(name) {
                        case 'rooms.json':
                            const rooms = filterFurnitures(data.rooms)
                            const furnitures = extractFurnitures(data.rooms, interactionData)
                            dispatch(loadRooms(rooms))
                            dispatch(loadFurnitures(furnitures))
                            break
                        case 'items.json':
                            dispatch(loadItems(data))
                            break
                        default:
                            console.log('Unrecognized file:',name)
                    }
                })
            })
        })
    }
}
