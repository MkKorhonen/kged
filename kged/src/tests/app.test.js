import React, { Component } from 'react'
import ReactDOM from 'react-dom'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import renderer from 'react-test-renderer'

import App from '../components/app'
import Rooms from '../components/rooms'
import CreateContainer from '../components/create_container'
import {addRoom} from '../actions'

const mockStore = configureStore([])

it('renders without crashing', () => {
  const div = document.createElement('div')
  const store = mockStore({
      rooms: {
          rooms: [],
          activeRoom: {}
      }
  })

  ReactDOM.render(
      <Provider store={store}>
        <App />
      </Provider>,
      div)
  ReactDOM.unmountComponentAtNode(div)
})

describe('Room component', () => {
  let store;
  let component;

  beforeEach(() =>{
    store = mockStore({
      rooms: {
        rooms: [],
        activeRoom: {}
      }
    });

    store.dispatch = jest.fn();

    component = renderer.create(
      <Provider store={store}>
        <Rooms />
      </Provider>
    );
  });

it('should open the creation container once clicked', () => {
  renderer.act(() => {
    component.root.findByProps({className:'my-3 btn btn-success'}).props.onClick();  });

    expect(CreateContainer.isActive = true);
  });
  it('should set a room active', () => {
    renderer.act(() => {
      component.root.findByProps({className:'room'}).props.onClick();

      expect(store.dispatch).toHaveBeenCalledTimes(1);
    });
  });
  it('should give the room a name and create it', () => {
    renderer.act(() => {
      component.root.findByType('name').props.onChange({target: {value: 'huone1'}});
      component.root.findByProps({className:'mr-2 btn btn-success'}).props.onClick();
      expect(store.dispatch).toHaveBeenCalledTimes(1);
      expect(store.dispatch).toHaveBeenCalledWith(addRoom({payload: 'huone1'}));
    });
  });
  
});

