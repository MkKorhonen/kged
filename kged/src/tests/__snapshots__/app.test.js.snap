// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Room component should render with given state from Redux store 1`] = `
<div>
  <div>
    <div>
      <button
        className="my-3 btn btn-success"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Lisää
      </button>
    </div>
  </div>
</div>
`;
